### PR TITLE
rsx/overlays: Reimplement message queue

### PR DESCRIPTION
**Pull request by kd-11**
Bunch of changes, both functional and aesthetic.
1. I had some code lying around for rounded rects, I made the overlay messages use this as well now before adding to other parts.
2. Implement message stacking. Don't display the same message multiple times.
3. Add both top pinned and bottom pinned notifications. Each set can display upto 3 simultaneous notifications. Notifications can be multiline as well without overlap issues.
4. Removes the shader compilation notification and replaces it with a notification message.
5. Add support for animated icons to the overlay system.
6. Add an animated icon to the shader compilation notification instead of the janky dots. Animation was done in gimp and exported and there is full support for sprite animation now in the UI. If someone has a better sprite-set, I can use it instead.
7. Messages have a seperate fade-in and fade-out animation as well. Makes messages spawning in less janky.

Some screenshots:
image
image
image

Known issues:
- Rounded edges with small radius look a bit janky due to some issues with shading. The meshes are generated correctly though, it's a polygon filling issue afaict.

Fixes https://github.com/RPCS3/rpcs3/issues/13270

### CODE REVIEW

**Comment by kd-11** [comment]
Linux build is borked, I'll have to check it tomorrow.

File: rpcs3/Emu/RSX/Overlays/overlay_message.h [hunk 1]
```diff
@@ -34,32 +47,64 @@ namespace rsx
...
 			shared_mutex m_mutex_queue;
-			std::deque<message_item> m_queue;
+
+			// Top and bottom enqueued sets
+			std::deque<message_item> m_ready_queue_top;
+			std::deque<message_item> m_ready_queue_bottom;
+
+			// Top and bottom visible sets
+			std::deque<message_item> m_vis_items_top;
+			std::deque<message_item> m_vis_items_bottom;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.h:91]: could just spell it out.

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 2]
```diff
@@ -101,13 +191,49 @@ namespace rsx
...
+			};
+
+			switch (location)
+			{
+			case message_pin_location::top:
+				return check_list(m_ready_queue_top) || check_list(m_vis_items_top);
+			case message_pin_location::bottom:
+				return check_list(m_ready_queue_bottom) || check_list(m_vis_items_bottom);
+			default:
+				fmt::throw_exception("Unreachable");
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:234]: if you remove this we'll get a warning if we forget to add a new location here.

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 3]
```diff
@@ -6,88 +6,178 @@ namespace rsx
...
+			{
+				m_expiration_time = get_expiration_time(m_visible_duration);
 			}
 
-			if ((m_expiration_time - time) < 2'000'000)
+			if (m_fade_in_animation.active)
+			{
+				m_fade_in_animation.update(rsx::get_current_renderer()->vblank_count);
+			}
+			else if (time + 2'000'000 > m_expiration_time)
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:123]: you can probably use the m_fade_out_animation.duration to calculate this

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 4]
```diff
@@ -6,88 +6,178 @@ namespace rsx
...
-			while (!m_queue.empty() && m_queue.front().get_expiration() < cur_time)
+			while (vis_set.size() < max_visible_items && !ready_set.empty())
 			{
-				m_queue.pop_front();
+				vis_set.emplace_back(std::move(ready_set.front()));
+				ready_set.pop_front();
 			}
 
-			if (m_queue.empty())
+			if (vis_set.empty() && ready_set.empty())
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:146]: do not check ready_set here

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 5]
```diff
@@ -6,88 +6,178 @@ namespace rsx
...
 			{
-				visible = false;
 				return;
 			}
 
-			usz index = 0;
+			// Render reversed list. Oldest entries are furthest from the border
+			u16 y_offset = 8, margin = 4, index = 0;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:152]: margin can be constexpr

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 6]
```diff
@@ -6,88 +6,178 @@ namespace rsx
...
+			for (auto it = vis_set.rbegin(); it != vis_set.rend(); ++it, ++index)
+			{
+				if (origin == message_pin_location::top) [[ likely ]]
+				{
+					it->update(index, cur_time, y_offset);
+					y_offset += (margin + it->h);
+				}
+				else
+				{
+					y_offset += (margin + it->h);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:162]: why do we add the margin beforehand? we already have the 8 pixel offset at the first message.

**Review by Megamouse — COMMENTED** [review]
It's odd that the upper margin of the text is always bigger.
I noticed this in my settings branch. I don't remember if i just added some offset or fixed it otherwise though.

File: rpcs3/Emu/RSX/Overlays/overlay_shader_compile_notification.h [hunk 7]
```diff
@@ -1,29 +1,21 @@
...
 #include "overlays.h"
+#include "overlay_message.h"
+#include "overlay_loading_icon.hpp"
 
 namespace rsx
 {
 	namespace overlays
 	{
-		struct shader_compile_notification : user_interface
+		static inline void show_shader_compile_notification()
```
[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/Overlays/overlay_shader_compile_notification.h:11]: idk, just felt like a waste of loc, but alright

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp (hunk 6 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/Overlays/overlay_message.cpp:162]: Looks nicer with a 16px offset vs 8. The same config up top looks awful though.

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp (hunk 2 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/Overlays/overlay_message.cpp:234]: image

Comment (kd-11) [reply rpcs3/Emu/RSX/Overlays/overlay_message.cpp]: Added a return false, though should really throw. We don't enforce warnings as error so many warning go unnoticed.

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp (hunk 3 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/Overlays/overlay_message.cpp]: took the opportunity to decrease fadeout to 1s as well

**Pull request merged**